lib/clusterable/ht_item.rb

```diff
@@ -4,56 +4,52 @@
 require "services"
 
 module Clusterable
-  # An HT Item
-  # - ocns
-  # - item_id
-  # - ht_bib_key
-  # - rights
-  # - bib_fmt
-  # - enum_chron
-  # - n_enum
-  # - n_chron
-  # - collection_code
-  # - billing_entity
   class HtItem
     include EnumChron
-    # field :ocns, type: Array, default: []
-    # field :item_id, type: String
-    # field :ht_bib_key, type: Integer
-    # field :rights, type: String
-    # field :access, type: String
-    # field :bib_fmt, type: String
-    # field :enum_chron, type: String, default: ""
-    # field :n_enum, type: String, default: ""
-    # field :n_chron, type: String, default: ""
-    # field :n_enum_chron, type: String, default: ""
-    # field :collection_code, type: String
-    # field :billing_entity, type: String
-
-    # embedded_in :cluster
-    # validates :item_id, uniqueness: true
-    # validates_presence_of :item_id, :ht_bib_key, :rights, :bib_fmt, :access
-
-    # validates_each :ocns do |record, attr, value|
-    #   value.each do |ocn|
-    #     record.errors.add attr, "must be an integer" \
-    #       unless (ocn.to_i if /\A[+-]?\d+\Z/.match?(ocn.to_s))
-    #   end
-    # end
-
-    def initialize(params = nil)
-      raise "not implemented"
-      super
-      set_billing_entity if collection_code
+
+    def self.accessor_attrs
+      [:ocns, :item_id, :ht_bib_key, :rights, :access, :bib_fmt, :n_enum, :n_chron, :n_enum_chron, :billing_entity, :enum_chron]
+    end
+
+    def self.all_attrs
+      accessor_attrs + [:collection_code]
+    end
+
+    accessor_attrs.each { |attr| attr_accessor attr }
+    attr_reader :collection_code
+
+    def self.find(item_id:)
+      row = Services.hathifiles_table.where(htid: item_id).first
+
```

Comment:
Do we want to raise if number of returned rows != 1? E.g.

```ruby
results = Services.hathifiles_table.where(htid: item_id)
raise KeyError, "Too many hits (#{results.size}) for item_id:'#{item_id}'" if results.size > 1
row = results.first
raise KeyError, "No htitem with key '#{item_id}'" unless row
```

Reply:
I think we can think of `Services.hathifiles_table` as a sort of API where we are guaranteed to receive 1 or 0 records, since `htid` is a primary key -- if that's not met, that's a problem with the API, not with our code, and I think checking it feels a bit unnecessarily defensive to me.

However, it does look like Sequel has a `single_record!` call that will raise an exception if there is no matching row  -- but it doesn't check if there are other rows. I think it does make the **intention** a little bit more clear than `first`. The documentation (http://sequel.jeremyevans.net/rdoc/classes/Sequel/Dataset.html#method-i-single_record-21) is a bit noncommittal here, but it probably doesn't hurt to try it.